coordinator/src/notifications.rs

```diff
@@ -0,0 +1,355 @@
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            NotificationKind::ChannelClose => write!(f, "ChannelClose"),
+            NotificationKind::PositionSoonToExpire => write!(f, "PositionSoonToExpire"),
+            NotificationKind::PositionExpired => write!(f, "PositionExpired"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Notification {
+    pub user_fcm_token: FcmToken,
+    pub notification_kind: NotificationKind,
+}
+
+impl Notification {
+    pub fn new(user_fcm_token: FcmToken, notification_kind: NotificationKind) -> Self {
+        Self {
+            notification_kind,
+            user_fcm_token,
+        }
+    }
+}
+
+/// Actor managing the notifications
+pub struct NotificationService {
+    notification_sender: mpsc::Sender<Notification>,
+}
+
+impl NotificationService {
+    /// Start the notification service
+    ///
+    /// If an empty string is passed in the constructor, the service will not send any notification.
+    /// It will only log the notification that it would have sent.
+    pub fn new(fcm_api_key: String) -> Self {
+        if fcm_api_key.is_empty() {
+            // Log it as error, as in production it should always be set
+            tracing::error!("FCM API key is empty. No notifications will not be sent.");
+        }
+
+        let (notification_sender, mut notification_receiver) = mpsc::channel(100);
+
+        // TODO: use RAII here
+        tokio::spawn({
+            let fcm_api_key = fcm_api_key;
+            let client = fcm::Client::new();
+            async move {
+                while let Some(Notification {
+                    user_fcm_token,
+                    notification_kind,
+                }) = notification_receiver.recv().await
+                {
+                    tracing::info!(%notification_kind, %user_fcm_token, "Sending notification");
+
+                    if !fcm_api_key.is_empty() {
+                        let notification = build_notification(notification_kind);
+                        if let Err(e) =
+                            send_notification(&client, &fcm_api_key, &user_fcm_token, notification)
+                                .await
+                        {
+                            tracing::error!("Could not send notification to FCM: {:?}", e);
+                        }
+                    }
+                }
+            }
+        });
+
+        Self {
+            notification_sender,
+        }
+    }
+
+    /// Constructs a new sender. Use a sender to send notification from any part of the system.
+    pub fn get_sender(&self) -> mpsc::Sender<Notification> {
+        self.notification_sender.clone()
+    }
+}
+
+/// Prepares the notification text
+fn build_notification<'a>(kind: NotificationKind) -> fcm::Notification<'a> {
+    let mut notification_builder = fcm::NotificationBuilder::new();
+    match kind {
+        NotificationKind::ChannelClose => {
```

Comment:
🔧 Can we please rename that notification kind to something like `PositionClose`, with `ChannelClose` I am automatically thinking of the LN Cannel.

Reply:
I've removed it completely! the idea to try to prompt the user when we would like to close the lightning LN channel collaboratively to avoid having locked up liquidity, but alas we're not using that, so there was little point in prematurely adding an enum variant!